core/src/main/scala/io/snappydata/Literals.scala

```diff
@@ -117,6 +117,8 @@ object Constant {
   val CHANGEABLE_JAR_NAME = "SNAPPY_CHANGEABLE_JAR_NAME"
 
   val RESERVOIR_AS_REGION = "spark.sql.aqp.reservoirAsRegion"
+
+  val FOLDABLE_FUNCTIONS = Set( "ROUND", "BROUND" )
```

Comment:
Other functions that depend on foldable:

- PERCENTILE (Percentile.scala)
- STACK (generators.scala)
- NTILE (windowExpressions.scala)
- STR_TO_MAP -- only 2nd & 3rd args, NAMED_STRUCT -- only odd position args (complexTypeCreator.scala)
- REFLECT and JAVA_METHOD -- only 1st and 2nd args (CallMethodViaReflection)
- XPATH, XPATH_BOOLEAN, ... all XPATH prefixed ones and only 2nd arg
- PERCENTILE_APPROX, APPROX_PERCENTILE -- only 2nd and 3rd args (ApproximatePercentile)

LIMIT argument tokenization should be turned off explicitly in Parser because it can come in a sub-query without a WHERE clause before it that would have turned tokenization off.
Same for SnappyParser.windowFrame

Tokenization for LIKE/RLIKE/REGEXP will work but will be inefficient since generated code will Pattern.compile the string for every row (as it assumes that expression can change). Since these are handled separately in SnappyParser so tokenization for only the right "termExpression" needs to be turned off.

The following functions will be more efficient on disabling and we don't expect constants to change:

- TRANSLATE (StringTranslate in stringExpressions.scala) -- ideally only 2nd and 3rd arguments
- UNIX_TIMESTAMP, TO_UNIX_TIMESTAMP, FROM_UNIXTIME, TO_UTC_TIMESTAMP, TRUNC, FROM_UTC_TIMESTAMP, NEXT_DAY (dateTimeExpressions.scala) -- ideally only 2nd argument
- LIKE, RLIKE -- only 2nd argument; seems like they can appear as just functions too as per FunctionRegistry

Selective disabling of tokenization for only required arguments will be better. The set of functions will become a map that will also list the arguments that have to disable tokenization and that can be used by parser to selectively replace instead of doing for "allExprs".

Also check if functions implementing CodegenFallback cause trouble (e.g. json_tuple) due to caching since those will ship Expression object and it could send back older one from the cached plan. If so change for all Expressions implementing CodegenFallback (e.g. using an Analyzer rule in SnappySessionState.analyzer).

Lastly two more points to track separately:

a) Right now ENABLE/DISABLE tokenization is a binary that turns on/off in the WHERE clause. If there is a sub-query in WHERE, for example, and then a further WHERE in sub-query, then tokenization will end at the end of the inner WHERE which could have gone on a bit further. A better fix for this is to turn it on globally for entire query and selectively turn it off only for clauses where it can cause trouble like being done for functions (only very few will remain like CAST dataType, ResolveInlineTables and few optimizations like RemoveLiteralFromGroupExpressions, EliminateSorts after rest of changes mentioned here are done).

b) ReorderAssociativeOperator is a rule to to fold all constants and we can track an issue to add similar for ParamLiterals for better performance. Likewise ConstantFolding will create new folded constant and equivalent rule for ParamLiteral can generate an initMutableState() with evaluated value instead of doing for each row. Similar for FoldablePropagation. Perhaps JVM will already optimize some of these?